exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java

```diff
@@ -306,14 +685,29 @@ public AggOutcome doWork() {
 
           while (true) {
             // Cleanup the previous batch since we are done processing it.
+            long pre = allocator.getAllocatedMemory();
             for (VectorWrapper<?> v : incoming) {
               v.getValueVector().clear();
             }
+            long beforeAlloc = allocator.getAllocatedMemory();
+
+            // Get the next RecordBatch from the incoming
             IterOutcome out = outgoing.next(0, incoming);
+
+            // If incoming batch is bigger than our estimate - adjust the estimate
+            long afterAlloc = allocator.getAllocatedMemory();
+            long incomingBatchSize = afterAlloc - beforeAlloc;
+            if ( /* ! handlingSpills && */ estMaxBatchSize < incomingBatchSize ) {
+               logger.trace("Found a bigger incoming batch: {} , prior estimate was: {}",incomingBatchSize,estMaxBatchSize);
+               estMaxBatchSize = incomingBatchSize;
+            }
+
             if (EXTRA_DEBUG_1) {
               logger.debug("Received IterOutcome of {}", out);
             }
             switch (out) {
+              case RESTART:
+                logger.warn("HASH AGG: doWork got a RESTART...");
```

Comment:
Shouldn't this throw an `IllegalStateException`? Something is terribly wrong if we get this where it is not expected. Simply logging zillions of messages to the log is not likely to call attention to the problem.